File: acoustic_sim/hat.py
import numpy as np
import scipy as sp

wave_length = sp.constants.speed_of_sound / 40_000
omega = 2 * np.pi * wave_length # angular frequency
k = 2 * np.pi / wave_length  # wave number

def run_hat(control_points, phase_res=16, z=0):
    # control_points = [np.array([0.025, 0.05, 0.1]), np.array([0.075, 0.05, 0.1])]
    control_points = np.asarray(control_points)
    tx, ty = np.meshgrid(np.linspace(0.005, 0.095, 10), np.linspace(0.005, 0.095, 10))
    transducers = np.stack([tx.flatten(), ty.flatten(), np.zeros([100]) + z], axis=1)
    # transducers = np.append(transducers, np.stack([tx.flatten(), ty.flatten(), np.zeros([100]) - z], axis=1), axis=0)

    # unsure why it needs to be the conjugate but do it for now
    # return np.conjugate(calc_transducer_phases(transducers, control_points, phase_res=phase_res))
    return calc_transducer_phases(transducers, control_points, phase_res=phase_res)

# far field piston-source model: from https://jontallen.ece.illinois.edu/uploads/473.F18/Lectures/Chapter_7b.pdf
def p(r, theta, t):
    p_0 = 1.293 # density of air
    U_0 = 1 # intensity, specifically, the speed at which the piston moves
    a = 0.005 # 5 mm: radius of the emitter
    if np.sin(theta) == 0: # avoid a divide by 0
        return 1j * omega * p_0 * a**2 * U_0 / (2 * r) * np.e**(1j * (omega * t + k * r))
    else:
        return 1j * omega * p_0 * a**2 * U_0 / (2 * r) * np.e**(1j * (omega * t + k * r)) * 2 * sp.special.j1(k * a * np.sin(theta)) / ( k * a * np.sin(theta))

def gen_propagators(transducers, control_points):
    # create propagators
    propagators = np.zeros([len(control_points), len(transducers)], dtype=complex)
    for i, cp in enumerate(control_points):
        for j, transducer in enumerate(transducers):
            vec_r = cp - transducer
            r = np.linalg.norm(vec_r)
            theta = np.arccos(vec_r[2]/r)
            propagators[i, j] = p(r, theta, 0)
    return propagators

def calc_transducer_phases(transducers, control_points, phase_res=16):
    propagators = gen_propagators(transducers, control_points)
    reflected_propagators = gen_propagators(transducers * [1, 1, -1], control_points)
    c_pressures = np.zeros([len(control_points)], dtype=complex)
    t_pressures = np.ones(len(transducers), dtype=complex)
    for _ in range(10):
        for i in range(len(c_pressures)):
            # forward propagate
            c = 0
            # direct contributions of transducers
            for j in range(len(t_pressures)):
                c += t_pressures[j] * propagators[i, j]

            # reflection contributions of transducers
            for j in range(len(t_pressures)):
                c += t_pressures[j] * reflected_propagators[i, j]
            # each control point has an amplitude of 1 / n
            c_pressures[i] = c / abs(c) * 1 / len(c_pressures)
            
        # backwards propagate
        for j in range(len(t_pressures)):
            pl = 0
            # direct contributions of transducers
            for i in range(len(c_pressures)):
                pl += c_pressures[i] * np.conjugate(propagators[i, j])

            # reflection contributions of transducers
            for i in range(len(c_pressures)):
                pl += c_pressures[i] * np.conjugate(reflected_propagators[i, j])

            t_pressures[j] = pl

        # normalize
        t_pressures = t_pressures / np.max(np.abs(t_pressures))

        # quantize  
        phase = np.round(np.angle(t_pressures) / (2*np.pi) * phase_res) / phase_res * 2 * np.pi
        amp = 1
    
        t_pressures = amp * np.e**(1j*phase)

    # TODO: make this not hard-coded
    return t_pressures.reshape([10, 10])

File: acoustic_sim/rust/src/hat.rs
use num::complex::Complex;
use std::f32::consts::PI;

use scilib::math::bessel;

const WAVE_LENGTH: f32 = 343.0 / 40000.0;
const OMEGA: f32 = 2.0 * PI * WAVE_LENGTH;
const K: f32 = 2.0 * PI / WAVE_LENGTH;

const P_0: f32 = 1.293; // density of air
const EMITTER_RADIUS: f32 = 0.005; // radius of the emitter: 5 mm

#[derive(Debug)]
pub struct Vec2D<T> {
    vec: Vec<T>,
    len_i: usize,
    len_j: usize,
}

impl<T: std::clone::Clone> Vec2D<T> {
    /// create a new Vec2D initialized to 0
    pub fn new(init: T, i: usize, j: usize) -> Vec2D<T> {
        Vec2D {
            len_i: i,
            len_j: j,
            vec: vec![init; i * j],
        }
    }

    pub fn set(&mut self, i: usize, j: usize, val: T) {
        self.vec[self.len_j * i + j] = val; // row-major order, row-first indexing
    }

    pub fn ix(&self, i: usize, j: usize) -> &T {
        &self.vec[self.len_j * i + j] // row-major order, row-first indexing
    }

    pub fn size(&self) -> (usize, usize) {
        (self.len_i, self.len_j)
    }
}

#[derive(Debug)]
pub struct Point {
    pub x: f32,
    pub y: f32,
    pub z: f32,
}

impl Point {
    pub fn add(&self, other: &Point) -> Point {
        Point {
            x: self.x + other.x,
            y: self.y + other.y,
            z: self.z + other.z,
        }
    }

    pub fn sub(&self, other: &Point) -> Point {
        Point {
            x: self.x - other.x,
            y: self.y - other.y,
            z: self.z - other.z,
        }
    }

    pub fn norm(&self) -> f32 {
        (self.x.powi(2) + self.y.powi(2) + self.z.powi(2)).sqrt()
    }
}

// TODO: add support for different transducer arrangements
pub fn run_hat(control_points: &Vec<Point>, phase_res: f32, z: f32) -> Vec<Complex<f32>> {
    let mut transducers: Vec<Point> = vec![];
    let sep = 0.01;
    let size = 10;

    for i in 0..size {
        for j in 0..size {
            transducers.push(Point {
                x: EMITTER_RADIUS + i as f32 * sep,
                y: EMITTER_RADIUS + j as f32 * sep,
                z: z,
            })
        }
    }

    return calc_transducer_phases(&transducers, control_points, phase_res);
}

// far field piston-source model: from https://jontallen.ece.illinois.edu/uploads/473.F18/Lectures/Chapter_7b.pdf
fn p(r: f32, theta: f32, t: f32) -> Complex<f32> {
    let sin_theta = f32::sin(theta);

    let amp: Complex<f32> = Complex::new(0.0, 1.0) * OMEGA * P_0 * EMITTER_RADIUS.powi(2)
        / (2.0 * r)
        * Complex::new(0.0, OMEGA * t + K * r).exp();

    if sin_theta == 0.0 {
        return amp;
    } else {
        return amp * 2.0 * bessel::j_n(1, (K * EMITTER_RADIUS * sin_theta) as f64) as f32
            / (K * EMITTER_RADIUS * sin_theta);
    };
}

fn gen_propagators(transducers: &Vec<Point>, control_points: &Vec<Point>) -> Vec2D<Complex<f32>> {
    let mut propagators: Vec2D<Complex<f32>> = Vec2D::new(
        Complex::new(0.0, 0.0),
        control_points.len(),
        transducers.len(),
    );

    for i in 0..control_points.len() {
        for j in 0..transducers.len() {
            let vec_r = control_points[i].sub(&transducers[j]);
            let r = vec_r.norm();
            let theta = (vec_r.z / r).acos();
            propagators.set(i, j, p(r, theta, 0.0));
        }
    }

    return propagators;
}

fn calc_transducer_phases(
    transducers: &Vec<Point>,
    control_points: &Vec<Point>,
    phase_res: f32,
) -> Vec<Complex<f32>> {
    let propagators = gen_propagators(transducers, control_points);
    let reflected_transducers: Vec<Point> = transducers
        .iter()
        .map(|p| Point {
            x: p.x,
            y: p.y,
            z: -p.z,
        })
        .collect();
    let reflected_propagators = gen_propagators(&reflected_transducers, control_points);
    let mut c_pressures = vec![Complex::<f32>::new(0.0, 0.0); control_points.len()];
    let mut t_pressures = vec![Complex::<f32>::new(1.0, 0.0); transducers.len()];

    for _iter in 0..10 {
        // forward propagate
        for i in 0..c_pressures.len() {
            let mut c = Complex::new(0.0, 0.0);

            // direct contributions
            for j in 0..t_pressures.len() {
                c += t_pressures[i] * propagators.ix(i, j);
            }

            // reflection contributions
            // TODO: can possibly add a reflection coefficient here to account for imperfect
            // reflective surface
            for j in 0..t_pressures.len() {
                c = c + t_pressures[i] * reflected_propagators.ix(i, j);
            }

            // each control point has an amplitude of 1 / n
            c_pressures[i] = c / c.norm() / c_pressures.len() as f32;
        }

        // backwards propagate
        for j in 0..t_pressures.len() {
            let mut pl: Complex<f32> = Complex::new(0.0, 0.0);

            // direct contributions
            for i in 0..c_pressures.len() {
                pl = pl + c_pressures[i] * propagators.ix(i, j).conj();
            }

            // reflection contributions
            // TODO: can possibly add a reflection coefficient here to account for imperfect
            // reflective surface
            for i in 0..c_pressures.len() {
                pl = pl + c_pressures[i] * reflected_propagators.ix(i, j).conj();
            }

            t_pressures[j] = pl;
        }

        // normalize
        let max = t_pressures
            .iter()
            .map(|p| p.norm())
            .max_by(|a, b| a.total_cmp(b))
            .unwrap();

        for p in &mut t_pressures {
            *p = *p / max;
        }

        // quantize
        for p in &mut t_pressures {
            let (_r, mut theta) = p.to_polar();
            theta = (theta / (2.0 * PI) * phase_res).round() * 2.0 * PI / phase_res;
            *p = Complex::from_polar(1.0, theta);
        }
    }

    return t_pressures;
}
